storage/transfermanager/downloader.go

```diff
@@ -17,61 +17,131 @@ package transfermanager
+	done      chan bool                 // Indicates to finish up work; expecting no more inputs.
+	workers   *sync.WaitGroup           // Keeps track of the workers that are currently running.
 }
 
-// DownloadObject queues the download of a single object. If it's larger than
-// the specified part size, the download will automatically be broken up into
-// multiple range reads. This will initiate the download but is non-blocking;
-// call Downloader.Results to process the result.
+// DownloadObject queues the download of a single object. This will initiate the
+// download but is non-blocking; call Downloader.Results to process the result.
 func (d *Downloader) DownloadObject(ctx context.Context, input *DownloadObjectInput) {
 	input.ctx = ctx
-	d.work <- input
+	d.addInput(input)
 }
 
-// Download a set of objects to a local path. Downloader will create the needed
-// directory structure locally as the operations progress.
-// This will initiate the download but is non-blocking; wait on Downloader.Results
-// to process the result. Results will be split into individual objects.
-// NOTE: Do not use, DownloadDirectory is not implemented.
-func (d *Downloader) DownloadDirectory(ctx context.Context, input *DownloadDirectoryInput) {
-	d.output <- &DownloadOutput{Bucket: input.Bucket, Err: errors.New("DownloadDirectory is not implemented")}
-	// This does obj listing/os.File calls and then converts to object inputs.
+// DownloadObjectWithCallback queues the download of a single object. This will
+// initiate the download but is non-blocking; use the callback to process the
+// result.
+// The results of downloads initiated with this method will not be provided in
+// the results given in Downloader.Results.
+func (d *Downloader) DownloadObjectWithCallback(ctx context.Context, input *DownloadObjectInput, callback func(*DownloadOutput)) {
+	input.ctx = ctx
+	input.callback = &callback
+	d.addInput(input)
 }
 
-// Waits for all outstanding downloads to complete. The Downloader must not be
-// used to download more objects or directories after this has been called.
+// WaitAndClose waits for all outstanding downloads to complete. The Downloader
+// must not be used for any more downloads after this has been called.
+// WaitAndClose returns an error wrapping all errors that were encountered by
+// the Downloader when downloading objects. These errors are also returned in
+// the DownloadOutput for the failing download.
 func (d *Downloader) WaitAndClose() error {
-	close(d.work)
+	d.done <- true
 	d.workers.Wait()
-	close(d.output)
+
+	if len(d.errors) > 0 {
+		err := errors.Join(d.errors...)
+		return fmt.Errorf("transfermanager: at least one error encountered downloading objects:\n%w", err)
+	}
 	return nil
 }
 
-// Results returns the iterator for download outputs.
-func (d *Downloader) Results() *DownloadOutputIterator {
-	return &DownloadOutputIterator{
-		output: d.output,
+// Results returns all the results of the downloads completed since the last
```

Comment:
I think we could probably enforce that this can only be called after the full job is complete (given that we are keeping WaitAndClose as a single func). Or we could even just return the slice from WaitAndClose. What do you think?

Reply:
As discussed, I'll change this to return results directly in WaitAndClose